samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,743 @@
+        self._symbol_resolver.resources = self._resources
+        self._symbol_resolver.parameters = self._parameters
+
+        self.intrinsic_key_function_map = {
+            IntrinsicResolver.FN_JOIN: self.handle_fn_join,
+            IntrinsicResolver.FN_SPLIT: self.handle_fn_split,
+            IntrinsicResolver.FN_SUB: self.handle_fn_sub,
+            IntrinsicResolver.FN_SELECT: self.handle_fn_select,
+            IntrinsicResolver.FN_BASE64: self.handle_fn_base64,
+            IntrinsicResolver.FN_FIND_IN_MAP: self.handle_find_in_map,
+            IntrinsicResolver.FN_TRANSFORM: self.handle_fn_transform,
+            IntrinsicResolver.FN_GET_AZS: self.handle_fn_get_azs,
+            IntrinsicResolver.REF: self.handle_fn_ref,
+            IntrinsicResolver.FN_GET_ATT: self.handle_fn_getatt,
+            IntrinsicResolver.FN_IMPORT_VALUE: self.handle_fn_import_value
+        }
+
+        self.conditional_key_function_map = {
+            IntrinsicResolver.FN_AND: self.handle_fn_and,
+            IntrinsicResolver.FN_OR: self.handle_fn_or,
+            IntrinsicResolver.FN_IF: self.handle_fn_if,
+            IntrinsicResolver.FN_EQUALS: self.handle_fn_equals,
+            IntrinsicResolver.FN_NOT: self.handle_fn_not
+        }
+
+    def intrinsic_property_resolver(self, intrinsic, parent_function="template"):
+        """
+        This resolves the intrinsic of the format
+        {
+            intrinsic: dict
+        } by calling the function with the relevant intrinsic function resolver.
+
+        This also supports returning a string, list, boolean, int since they may be intermediate steps in the recursion
+        process. No transformations are done on these.
+
+        By default this will just return the item if non of the types match. This is because of the function
+        resolve_all_attributes which will recreate the resources by processing every aspect of resource.
+
+        Parameters
+        ----------
+        intrinsic: dict, str, list, bool, int
+            This is an intrinsic property or an intermediate step
+        parent_function: str
+            In case there is a missing property, this is used to figure out where the property resolved is missing.
+
+        Return
+        ---------
+        The simplified version of the intrinsic function. This could be a list,str,dict depending on the format required
+        """
+        if intrinsic is None:
+            raise InvalidIntrinsicException("Missing Intrinsic property in {}".format(parent_function))
+
+        if any(isinstance(intrinsic, object_type) for object_type in [string_types, list, bool, int]):
+            return intrinsic
+
+        if intrinsic == {}:
+            return intrinsic
+
+        keys = list(intrinsic.keys())
+        key = keys[0]
+
+        if key in self.intrinsic_key_function_map:
+            intrinsic_value = intrinsic.get(key)
+            return self.intrinsic_key_function_map.get(key)(intrinsic_value)
```

Comment:
If we have intrinsic inside intrinsic, it is resolved within each specific intrinisic_function. This is because there are special cases to consider for each type when dealing with functions within functions. For example, Fn::Select needs to have the first element be an integer index.